cache/pmdk_cache.cc

```diff
@@ -0,0 +1,369 @@
+  size_t total_charge = e->CalcTotalCharge(metadata_charge_policy_);
+  if (high_pri_pool_ratio_ > 0 && (e->IsHighPri() || e->HasHit())) {
+    // Inset "e" to head of LRU list.
+    e->next = &lru_;
+    e->prev = lru_.prev;
+    e->prev->next = e;
+    e->next->prev = e;
+    e->SetInHighPriPool(true);
+    high_pri_pool_usage_ += total_charge;
+    MaintainPoolSize();
+  } else {
+    // Insert "e" to the head of low-pri pool. Note that when
+    // high_pri_pool_ratio is 0, head of low-pri pool is also head of LRU list.
+    e->next = lru_low_pri_->next;
+    e->prev = lru_low_pri_;
+    e->prev->next = e;
+    e->next->prev = e;
+    e->SetInHighPriPool(false);
+    lru_low_pri_ = e;
+  }
+  lru_usage_ += total_charge;
+}
+
+void PMDKCacheShard::MaintainPoolSize() {
+  while (high_pri_pool_usage_ > high_pri_pool_capacity_) {
+    // Overflow last entry in high-pri pool to low-pri pool.
+    lru_low_pri_ = lru_low_pri_->next;
+    assert(lru_low_pri_ != &lru_);
+    lru_low_pri_->SetInHighPriPool(false);
+    size_t total_charge =
+        lru_low_pri_->CalcTotalCharge(metadata_charge_policy_);
+    assert(high_pri_pool_usage_ >= total_charge);
+    high_pri_pool_usage_ -= total_charge;
+  }
+}
+
+void PMDKCacheShard::EvictFromLRU(size_t charge,
+                                 autovector<TransientHandle*>* deleted) {
+  // TODO
+}
+
+void PMDKCacheShard::SetCapacity(size_t capacity) {
+  autovector<TransientHandle*> last_reference_list;
+  {
+    MutexLock l(&mutex_);
+    capacity_ = capacity;
+    high_pri_pool_capacity_ = capacity_ * high_pri_pool_ratio_;
+    EvictFromLRU(0, &last_reference_list);
+  }
+
+  // Free the entries outside of mutex for performance reasons
+  for (auto entry : last_reference_list) {
+    entry->Free();
+  }
+}
+
+void PMDKCacheShard::SetStrictCapacityLimit(bool strict_capacity_limit) {
+  MutexLock l(&mutex_);
+  strict_capacity_limit_ = strict_capacity_limit;
+}
+
+Cache::Handle* PMDKCacheShard::Lookup(const Slice& key, uint32_t hash,
+                                      void* (*pack)(const Slice& slice)) {
+  // TODO: lookup transient hash table
+  // lookup transient table:
+  TransientHandle* e = persistent_hashtable_->Lookup(key, hash);
+  // TODO: LRU operations.
+  return reinterpret_cast<Cache::Handle*>(e);
+}
+
+bool PMDKCacheShard::Ref(Cache::Handle* h) {
+  TransientHandle* e = reinterpret_cast<TransientHandle*>(h);
+  MutexLock l(&mutex_);
+  // To create another reference - entry must be already externally referenced
+  assert(e->HasRefs());
+  e->Ref();
+  return true;
+}
+
+void PMDKCacheShard::SetHighPriorityPoolRatio(double high_pri_pool_ratio) {
+  MutexLock l(&mutex_);
+  high_pri_pool_ratio_ = high_pri_pool_ratio;
+  high_pri_pool_capacity_ = capacity_ * high_pri_pool_ratio_;
+  MaintainPoolSize();
+}
+
+bool PMDKCacheShard::Release(Cache::Handle* handle, bool force_erase) {
+  // TODO
+  return true;
+}
+
+Status PMDKCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
```

Comment:
I think things here can be more complex. Assume `value` points to a `Block` or `BlockContents`. In `LRUCache`, if insertion (to volatile cache) fails, then this function returns a non-ok status to caller. The caller is expected to release the buffer pointed to by `BlockContents::allocation`.
In comparison, if `PMDKCache` fails to insert to volatile tier, but succeeds in inserting to persistent tier. Then we should release the memory in volatile memory. This means the `BlockContents::allocation` will need to be changed to point to persistent memory. I haven't figured this out yet.

Reply:
Maybe get rid of volatile tier, but keep the volatile memory allocator?

Reply:
We have deleter passed in, right? Is that enough to get rid of the transient part?

Reply:
If there is ever a hit in persistent tier, the cache is responsible for re-building `BlockContents`. Maybe that's where we can put persistent memory management as `BlockContents::allocation`?

Reply:
You can use the deleter to free volatile memory, but `value` will need to be updated so that it (its member variable) points to persistent memory. It's like you want to put something in the cache, but you end up deleting what you already have and caching something else.
I am thinking about how `BlockBasedTable` can use this cache.

Reply:
We can free it after inserting into persistent tier, perhaps.

Reply:
Also remember that, for uncompressed block cache, the allocator will be used to allocate space for uncompressed data in addition to the previously-allocated io buffer.